compress.cc

```diff
@@ -531,3 +652,391 @@ size_t snappy_processor::compress_max_size(size_t input_len) const {
+    case algorithm::zstd: {
+        co_return seastar::make_shared<zstd_processor>(params);
+    }
+    case algorithm::none:
+        co_return nullptr;
+    }
+    abort();
+}
+
+future<compressor_ptr> sstable_compressor_factory_impl::make_compressor_for_reading(sstables::compression& c) {
+    auto params = compression_parameters(sstables::options_from_compression(c));
+    using algorithm = compression_parameters::algorithm;
+    switch (params.get_algorithm()) {
+    case algorithm::lz4:
+        co_return compressor::lz4;
+    case algorithm::deflate:
+        co_return compressor::deflate;
+    case algorithm::snappy:
+        co_return compressor::snappy;
+    case algorithm::zstd:
+        co_return seastar::make_shared<zstd_processor>(params);
+    case algorithm::none:
+        co_return nullptr;
+    }
+}
+
+raw_dict::raw_dict(sstable_compressor_factory_impl& owner, dict_id key, std::span<const std::byte> dict)
+    : _owner(&owner)
+    , _id(key)
+    , _dict(dict.begin(), dict.end())
+{
+    _owner->account_memory_delta(malloc_usable_size(const_cast<std::byte*>(_dict.data())));
+}
+
+raw_dict::~raw_dict() {
+    if (_owner) {
+        disown();
+    }
+}
+
+void raw_dict::disown() const noexcept {
+    _owner->forget_raw_dict(_id);
+    _owner->account_memory_delta(-malloc_usable_size(const_cast<std::byte*>(_dict.data())));
+    _owner = nullptr;
+}
+
+zstd_cdict::zstd_cdict(sstable_compressor_factory_impl& owner, lw_shared_ptr<const raw_dict> raw, int level)
+    : _owner(&owner)
+    , _raw(raw)
+    , _level(level)
+    , _alloc([this] (ssize_t n) {
+        _used_memory += n;
+        if (_owner) {
+            _owner->account_memory_delta(n);
+        }})
+    , _dict(
+        ZSTD_createCDict_advanced(
+            _raw->raw().data(),
+            _raw->raw().size(),
+            ZSTD_dlm_byRef,
+            ZSTD_dct_auto,
+            ZSTD_getCParams(level, 4096, _raw->raw().size()),
+            _alloc.as_zstd_custommem()),
+        ZSTD_freeCDict)
+{
+    if (!_dict) {
+        throw std::bad_alloc();
+    }
+}
+
+
+zstd_cdict::~zstd_cdict() {
+    if (_owner) {
+        disown();
+    }
+}
+
+void zstd_cdict::disown() const noexcept {
+    _owner->forget_zstd_cdict(_raw->id(), _level);
+    _owner->account_memory_delta(-_used_memory);
+    _owner = nullptr;
+}
+
+zstd_ddict::zstd_ddict(sstable_compressor_factory_impl& owner, lw_shared_ptr<const raw_dict> raw)
+    : _owner(&owner)
+    , _raw(raw)
+    , _alloc([this] (ssize_t n) {
+        _used_memory += n;
+        if (_owner) {
+            _owner->account_memory_delta(n);
+        }})
+    , _dict(
+        ZSTD_createDDict_advanced(
+            _raw->raw().data(),
+            _raw->raw().size(),
+            ZSTD_dlm_byRef,
+            ZSTD_dct_auto,
+            _alloc.as_zstd_custommem()),
+        ZSTD_freeDDict)
+{
+    if (!_dict) {
+        throw std::bad_alloc();
+    }
+}
+
+zstd_ddict::~zstd_ddict() {
+    if (_owner) {
+        disown();
```

Comment:
Is this disown() stuff a reimplementation of weak_ptr?

Reply:
Yes, it is.

I had `weak_ptr` in mind and didn't use it for some conscious reason. But now that I look at the code, I don't see the reason anymore.

Will modify.

Reply:
Done.